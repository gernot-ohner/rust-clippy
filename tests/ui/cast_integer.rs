#![warn(clippy::cast_integer)]

fn main() {
    let u = 1usize;
    let _ = u as u64;
    let _ = u64::try_from(u);

    // unsigned
    let u64= 1u64;
    let _ = u64 as u128;
    let _ = u128::from(u64);

    let u32 = 1u32;
    let _ = u32 as u128;
    let _ = u128::from(u32);
    let _ = u32 as u64;
    let _ = u64::from(u32);

    let u16= 1u16;
    let _ = u16 as u128;
    let _ = u128::from(u16);
    let _ = u16 as u64;
    let _ = u64::from(u16);
    let _ = u16 as u32;
    let _ = u32::from(u16);

    let u8 = 1u8;
    let _ = u8 as u128;
    let _ = u128::from(u8);
    let _ = u8 as u64;
    let _ = u64::from(u8);
    let _ = u8 as u32;
    let _ = u32::from(u8);
    let _ = u8 as u16;
    let _ = u16::from(u8);

    // signed

    let isize= 1isize;
    let _ = isize as i64;
    let _ = i64::try_from(isize);

    let i64 = 1i64;
    let _ = i64 as i128;
    let _ = i128::from(i64);

    let i32 = 1i32;
    let _ = i32 as i128;
    let _ = i128::from(i32);
    let _ = i32 as i64;
    let _ = i64::from(i32);

    let i16 = 1i16;
    let _ = i16 as i128;
    let _ = i128::from(i16);
    let _ = i16 as i64;
    let _ = i64::from(i16);
    let _ = i16 as i32;
    let _ = i32::from(i16);

    let i8 = 1i8;
    let _ = i8 as i128;
    let _ = i128::from(i8);
    let _ = i8 as i64;
    let _ = i64::from(i8);
    let _ = i8 as i32;
    let _ = i32::from(i8);
    let _ = i8 as i16;
    let _ = i16::from(i8);
}

// The lint would suggest using `u32::from(input)` here but the `XX::from` function is not const,
// so we skip the lint if the expression is in a const fn.
// See #3656
const fn const_function(input: u32) -> u64 {
    input as u64
}

// Same as the above issue. We can't suggest `::from` in const fns in impls
mod cast_integer_in_impl {
    struct A;

    impl A {
        pub const fn convert(x: u32) -> u64 {
            x as u64
        }
    }
}