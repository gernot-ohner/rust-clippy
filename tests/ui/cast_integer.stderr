error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:5:13
   |
LL |     let _ = u as u64;
   |             ^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead
   = note: `-D clippy::cast-integer` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::cast_integer)]`

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:10:13
   |
LL |     let _ = u64 as u128;
   |             ^^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:14:13
   |
LL |     let _ = u32 as u128;
   |             ^^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:16:13
   |
LL |     let _ = u32 as u64;
   |             ^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:20:13
   |
LL |     let _ = u16 as u128;
   |             ^^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:22:13
   |
LL |     let _ = u16 as u64;
   |             ^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:24:13
   |
LL |     let _ = u16 as u32;
   |             ^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:28:13
   |
LL |     let _ = u8 as u128;
   |             ^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:30:13
   |
LL |     let _ = u8 as u64;
   |             ^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:32:13
   |
LL |     let _ = u8 as u32;
   |             ^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:34:13
   |
LL |     let _ = u8 as u16;
   |             ^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:40:13
   |
LL |     let _ = isize as i64;
   |             ^^^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:44:13
   |
LL |     let _ = i64 as i128;
   |             ^^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:48:13
   |
LL |     let _ = i32 as i128;
   |             ^^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:50:13
   |
LL |     let _ = i32 as i64;
   |             ^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:54:13
   |
LL |     let _ = i16 as i128;
   |             ^^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:56:13
   |
LL |     let _ = i16 as i64;
   |             ^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:58:13
   |
LL |     let _ = i16 as i32;
   |             ^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:62:13
   |
LL |     let _ = i8 as i128;
   |             ^^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:64:13
   |
LL |     let _ = i8 as i64;
   |             ^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:66:13
   |
LL |     let _ = i8 as i32;
   |             ^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: Integer casts can introduce subtle surprises and should be done with From/TryFrom.
  --> $DIR/cast_integer.rs:68:13
   |
LL |     let _ = i8 as i16;
   |             ^^^^^^^^^
   |
   = help: Try T::from(_) or T::try_from(_) instead

error: aborting due to 22 previous errors

